lib/linguist/languages.yml

```diff
@@ -4554,6 +4554,19 @@ Raw token data:
   tm_scope: none
   ace_mode: text
   language_id: 318
+ReScript:
+  type: programming
+  color: "#df4b37"
+  ace_mode: rust
+  codemirror_mode: rust
+  codemirror_mime_type: text/x-rustsrc
+  extensions:
+  - ".res"
```

Comment:
This extension is incredibly popular on GItHub.com. If we were to merge this PR as-is, they will _all_ be identified as ReScript. As such, if you want to add this extension, you will also need to try and identify the language of as many of the other non-ReScript `.res` files as possible, add them to this PR, and implement a heuristic to differentiate between them.

Reply:
Just for clarification: to be able to correctly detect the ReScript language, we need to add lots of example files with different ReScript code scenarios and add some regex rules to [heuristics.yml](https://github.com/github/linguist/blob/master/lib/linguist/heuristics.yml)?

Should we add non ReScript `.res` files in the examples folder as well and make sure those are not detected?

Reply:
> Just for clarification: to be able to correctly detect the ReScript language, we need to add lots of example files with different ReScript code scenarios and add some regex rules to [heuristics.yml](https://github.com/github/linguist/blob/master/lib/linguist/heuristics.yml)?

Not quite. It's not a case of only correctly identifying ReScript, it's about identifying all the other users too.

If we were to flip the scenario around... hypothetically,  if I beat you to it and added a new language called "Lildude's Cool Language" and associated the `.res` with it, every single ReScript file will be identified as "Lildude's Cool Language" because there is no other language associated with that extension. I'm pretty sure the ReScript users wouldn't be too pleased with this, hence we need to add as many of the other users of `.res` to Linguist at the same time so those users don't suddenly have all their files identified as ReScript.

So in other words, this PR will need to be for the addition of ReScript, _and_ "Language1 Using `.res`", _and_ "Language2 Using `.res`" etc.

Reply:
I just added an entry for `.res` in the `heuristics.yml` file, including a pattern for `ReScript`:

@ryyppy can you let me know if it's enough to cover rescript files?https://github.com/github/linguist/pull/4975/commits/74da5fd875c9a5729a4cdbc2b46c5c6dd27bffa6

Reply:
@lildude thanks for the clarification, I think I now understand what you mean.. I guess @brnrdog just started the right actions here :)

@brnrdog I think for the `heuristics.yml` we could also add this (basically like in Reason):

```diff
pattern: 
+- '^\s*module\s+type\s'
 - '^\s*(let|module|type).*='
+- '^\s*(?:include|open)\s+\w+\s*;\s*$'
```

Reply:
@ryyppy I updated the patterns, I've changed a little bit since ReScript does not require semi-colon for the statements, (correct me if I'm wrong): https://github.com/github/linguist/pull/4975/commits/5f61ff04844ea2c09092ab61194160aa09d16931

Reply:
ah right, yeah... semicolons are actually optional here, so leaving them out should be fine.

Not sure what else would be missing now. Do we still need to find colliding file types and disambiguate them separately? I looked through global Github search, and many `.res` files are wildly different plain text formats that sometimes resemble XML flavors, plain text or binary files etc... no idea if those would be classified wrongly, now that we have some heuristics?